### ISSUE TITLE
Sending RDNs in Reverse Order for IssuerAndSerial

### ISSUE THREAD

**Issue by seize-the-dave** [jscep/jscep#70]
From Kevin Wallis:

The following code yields error logs from OpenXPKI

```java
KeyPair jscepKeyPair = createRandomKeyPair();
X509Certificate jscepCertificate = createOwnCertificate(jscepKeyPair, client);
...
X500Principal crlIssuer = new X500Principal("CN=CA ONE,OU=Test CA,DC=OpenXPKI,DC=ORG");
BigInteger crlSerialNumber = BigInteger.valueOf(511);
X509CRL crl = client.getRevocationList(jscepCertificate, jscepKeyPair.getPrivate(), crlIssuer, crlSerialNumber);
```

Error log:

`openxpki.application.ERROR:31458 [OpenXPKI::Service::SCEP::Command::PKIOperation (/usr/lib/x86_64-linux-gnu/perl5/5.20/OpenXPKI/Service/SCEP/Command/PKIOperation.pm:255); scep-server-1()@2dec] SCEP getcrl - no issuer found for serial 511 and issuer DC=ORG,DC=OpenXPKI,OU=Test CA,CN=CA ONE`

Note the difference in order of the issuer (DC first vs DC last).  This issue will be used to determine if this is a jscep issue, or and issue with OpenXPKI.

**Comment by seize-the-dave** [comment]
`getRevocationList` calls `new X500Name(issuer.getName())`.  X500Name is encoded by BouncyCastle, so this is really out of our control at this point.

There are optional arguments to both `X500Principle.getName()` and the `X500Name` constructor.  The test case should be whether encoding and decoding the X500Name yields the same representation on the jscep side.

**Comment by seize-the-dave** [comment]
Created new branch for some test cases: https://github.com/jscep/jscep/tree/issue-70

**Comment by seize-the-dave** [comment]
Original message can be seen here: https://lapo.it/asn1js/#308006092A864886F70D010702A0803080020101310F300D06096086480165030402030500308006092A864886F70D010701A080248004820225308006092A864886F70D010703A080308002010031820174308201700201003058305331133011060A0992268993F22C64011916034F524731183016060A0992268993F22C64011916084F70656E58504B493110300E060355040B0C07546573742043413110300E06035504030C07526F6F74204341020102300D06092A864886F70D01010105000482010011A4C1C707CF0F2CA18BA594CC76A9AA5037B74741B9CFEA21EB70DC81461E217748B5F7834F720FDC44B714B2F02CCFB89ABA19920BAD9D24C707CA47906D57945CA1F8B145451C991CD2BA193580B6738CD2FD1EDF42EB92C42B2FC8570C6BFC4DD364C96E0376115D1B11582E8EB1CAD381574139BA78451B810AFE79A11B620A6BD933C40C364C62054038DCF94921B97A3BB74C76B7C64FDD5840A49FB257DBD6E349AD77AD9BDB9D5BEB5EFAD9A28AC5327F68DCAAFE45FA92C9BFC1537CF015708D76A36AFAD4F5BAB8542A1136B93005B084C0A3B4DB470D11588FE766BFB72F6877DC0777EC56D7EB456769EAE690D8F824DEA3B1A57D6BEE2C66C3308006092A864886F70D010701301406082A864886F70D03070408F52D6BE484AAD41FA0800468D4B9D1F9165BEB676C0173F4AFEA7A20BA0E17C5545A3DDD0FE3849A9AC4EE02D0706D36CC11E62340E5630C244D240907B51EE63047A5DA1CBE65F62C70C0EE216F2EC864C8940C3A9A62D6A01270E6E2AF77D0C9F9C82B361533B9E4B02F7286ABB91A0459F80900000000000000000000000000000000A0803082023C308201A5A003020102020101300D06092A864886F70D01010D0500306431133011060A0992268993F22C64011916036F726731183016060A0992268993F22C64011916084F70656E58504B49311F301D060A0992268993F22C640119160F54657374204465706C6F796D656E7431123010060355040313096A736365702E6F7267301E170D3137303531353135303934365A170D3137303531373135303934365A306431133011060A0992268993F22C64011916036F726731183016060A0992268993F22C64011916084F70656E58504B49311F301D060A0992268993F22C640119160F54657374204465706C6F796D656E7431123010060355040313096A736365702E6F726730819F300D06092A864886F70D010101050003818D0030818902818100ACA016705164DF4CEFCA1A8354629FCEB4F762BB42CE0D00AD0B4EA1A9FC1DFD1F6874252A7FAD7972E1616B28067BA155979B64E7A8D7146A0AAB0AAA33DED3841536A4CB8C259046962D2EA43659A0178291F2BCC2728BA664C8F435E852D270753413556E042223E97675BCB9FEF4FDF96D6F72E258FBD8A8D5F381298F5D0203010001300D06092A864886F70D01010D0500038181009A563C985C632B4ABDCEFBFEF0A23AB88280FAF62AC617C309CB2DBEC4B6849C3B116CAD2870B05388A5BA4163B87C5EF8BEB37E82A81D2BF1DEFECB06053AA8AD07440D01A5EC961243555461654BB8A2FBBBD00965D191089A3A3955439EEF794A8AB017A5EA11094FE38046D5C7FCB6E787489CB74B652F3D44D3E0B0DC0F0000318201E8308201E40201013069306431133011060A0992268993F22C64011916036F726731183016060A0992268993F22C64011916084F70656E58504B49311F301D060A0992268993F22C640119160F54657374204465706C6F796D656E7431123010060355040313096A736365702E6F7267020101300D06096086480165030402030500A081D23011060A6086480186F84501090731031301303012060A6086480186F845010902310413023232301806092A864886F70D010903310B06092A864886F70D010701301C06092A864886F70D010905310F170D3137303531363135303934375A3020060A6086480186F84501090531120410E961A2F7A53050099647FE0E50125D7C304F06092A864886F70D010904314204408FB0D309961900A74C86B7853591F40AB41C93787B93318C1280C16BCE7B917CD77F3E8B4D850E415E213566958F9E58717A1314C824F03B132F882D9384FED1300D06092A864886F70D0101010500048180207EED49BB5E26E9CED3A00BABEE2D17BBFC93A4D99E6C2BD238DAF6F3E83CADD636ABEE05C193B6EA8F2644467F4104C979324B6C8CD9543083EE4F09551686CCF24CC6F316A8394DCBD34792866DA6C8A425B91BD3C94BAE1E8147CB5D9660681ED497F48C17351C13AE4692DB5CF89BEB53BD3FBF48E4438B9ACF99B0B664000000000000

**Comment by netcat-pl** [comment]
> getRevocationList calls new X500Name(issuer.getName()). X500Name is encoded by BouncyCastle, so this is really out of our control at this point.

This isn't true that this error is caused by BouncyCastle. X500Principal.getName() returns DN string formated according to RFC2253 which means that all RDNs are formated in reverse order. And that is the case. You can't use this function to construct org.bouncycastle.asn1.x500.X500Name object. So the code in org.jscep.client.Client.getRevocationList():
```
X500Name name = new X500Name(issuer.getName());
IssuerAndSerialNumber iasn = new IssuerAndSerialNumber(name, serial);
```
And the code in org.jscep.client.Client.getCertificate():
```
X500Name name = new X500Name(ca.getSubjectX500Principal().toString());
IssuerAndSerialNumber iasn = new IssuerAndSerialNumber(name, serial);
```
Is wrong. Instead, you should use org.jscep.util.X500Utils.toX500Name() as in org.jscep.client.Client.poll() function:
```
IssuerAndSubject ias = new IssuerAndSubject(X500Utils.toX500Name(issuer
.getSubjectX500Principal()), X500Utils.toX500Name(subject));
```

**Comment by seize-the-dave** [comment]
@netcat-pl are you will to provide a pull request for this?

**Comment by seize-the-dave** [comment]
This is now available in release 2.5.1
